python/cusignal/io/reader.py

```diff
@@ -0,0 +1,308 @@
+        read in all samples.
+    offset : int, optional
+        In the file, array data starts at this offset.
+        Since offset is measured in bytes, it should normally
+        be a multiple of the byte-size of dtype.
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing binary data.
+
+    """
+
+    # Get current stream, default or not.
+    stream = cp.cuda.get_current_stream()
+
+    # offset is measured in bytes
+    offset *= cp.dtype(dtype).itemsize
+
+    fp = np.memmap(file, mode="r", offset=offset, shape=num_samples)
+
+    if buffer is not None:
+        out = cp.empty(buffer.shape, buffer.dtype)
+
+    if buffer is None:
+        out = cp.asarray(fp)
+    else:
+        buffer[:] = fp[:]
+        out.set(buffer)
+
+    stream.synchronize()
+
+    del fp
+
+    return out
+
+
+def write_bin(file, binary, buffer=None, append=True):
+    """
+    Writes binary array to file.
+
+    Parameters
+    ----------
+    file : str
+        A string of filename to store output.
+    binary : ndarray
+        Binary array to be written to file.
+    buffer : ndarray, optional
+        Pinned memory buffer to use when copying data from GPU.
+    append : bool, optional
+        Append to file if created.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing binary data.
+
+    """
+
+    # Get current stream, default or not.
+    stream = cp.cuda.get_current_stream()
+
+    if buffer is None:
+        buffer = cp.asnumpy(binary)
+    else:
+        binary.get(out=buffer)
+
+    if append is True:
+        mode = "ab"
+    else:
+        mode = "wb"
+
+    with open(file, mode) as f:
+        stream.synchronize()
+        buffer.tofile(f)
+
+
+def unpack_bin(binary, dtype, endianness="L"):
+    """
+    Unpack binary file.
+    If endianness is big-endian, it my be converted
+    to little endian for NVIDIA GPU compatibility.
+
+    Parameters
+    ----------
+    binary : ndarray
+        The binary array to be unpack.
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy data type.
+    endianness : {'L', 'B'}, optional
+        Data set byte order
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing unpacked binary data.
+
+    """
+
+    if endianness != "L" and endianness != "B" and endianness != "N":
+        raise ValueError("'endianness' should be 'L' or 'B'")
+
+    out = _unpack(binary, dtype, endianness)
+
+    return out
+
+
+def pack_bin(in1):
+    """
+    Pack binary arrary.
+    Data will be packed with little endian for NVIDIA GPU compatibility.
+
+    Parameters
+    ----------
+    in1 : ndarray
+        The ndarray to be pack at binary.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing packed binary data.
+
+    """
+
+    out = _pack(in1)
+
+    return out
+
+
+def read_sigmf(data_file, meta_file, buffer=None, num_samples=None, offset=0):
```

Comment:
I think this is a fine API, but I'm curious if we should make the metadata file optional and put the header read in an if statement.

Reply:
How about this?
```python
if meta_file is None:
        meta_ext = ".sigmf-meta"

        split_string = data_file.split(".")
        meta_file = split_string[0] + meta_ext

    with open(meta_file, "r") as f:
        header = json.loads(f.read())
```